### PR TITLE
Categorical Generator w Probabilities

### PR DESCRIPTION
**Pull request by lizlouise1335**
Cat generator now handles probabilities (tests too)

### CODE REVIEW

File: tests/distinct_generators/test_categorical_generator.py [hunk 1]
```diff
@@ -9,7 +9,7 @@
 class TestRandomsCategories(unittest.TestCase):
     def setUp(self):
         self.rng = Generator(PCG64(12345))
-        self.categories = ["People", "Cats", "Dogs"]
+        self.categories = ["People", "Cats", "Dogs", "Bees", "Spiders"]
```
Comment (taylorfturner) [inline tests/distinct_generators/test_categorical_generator.py:12]:
```suggestion
        self.categories = ["Lions", "Tigers", "Bears", "Oh my"]
```
missed opportunity.... sadness

Comment (taylorfturner) [reply tests/distinct_generators/test_categorical_generator.py:12]: goofin'

File: synthetic_data/distinct_generators/categorical_generator.py [hunk 2]
```diff
@@ -24,4 +27,11 @@ def random_categorical(
         categories = ["A", "B", "C", "D", "E"]
     if num_rows > len(categories):
         raise ValueError("num_rows exceeds number of categories")
-    return rng.choice(categories, (num_rows,))
+    if probabilities is None:
+        return rng.choice(categories, size=num_rows)
+
+    if len(categories) != len(probabilities):
+        raise ValueError("categories and probabilities must be of the same length")
```
Comment (taylorfturner) [inline synthetic_data/distinct_generators/categorical_generator.py:34]: should `assertRaisesRegex`

File: synthetic_data/distinct_generators/categorical_generator.py [hunk 3]
```diff
@@ -24,4 +27,11 @@ def random_categorical(
...
+    if not np.isclose(sum(probabilities), 1):
+        raise ValueError("Probabilities must sum to 1")
```
Comment (taylorfturner) [inline synthetic_data/distinct_generators/categorical_generator.py:36]: should `assetRaisesRegex`

Comment (taylorfturner) [reply synthetic_data/distinct_generators/categorical_generator.py:36]: validate in a test

File: tests/distinct_generators/test_categorical_generator.py [hunk 4]
```diff
@@ -27,3 +27,30 @@ def test_custom_return_validity(self):
...
+
+        cat_probs = [0.1, 0.4, 0.1, 0.3, 0.1]
+        for i in range(10):
+            result = random_categorical(
+                self.rng,
+                categories=self.categories,
+                num_rows=5,
+                probabilities=cat_probs,
+            )
+            print(result)
```
Comment (taylorfturner) [inline tests/distinct_generators/test_categorical_generator.py:53]: no bueno on the prints

File: tests/distinct_generators/test_categorical_generator.py [hunk 5]
```diff
@@ -27,3 +27,30 @@ def test_custom_return_validity(self):
...
+    def test_weighted_return_validity(self):
+        cat_probs = [0.1, 0.5, 0.0, 0.3, 0.1]
+        for i in range(10):
+            result = random_categorical(
+                self.rng,
+                categories=self.categories,
+                num_rows=5,
+                probabilities=cat_probs,
+            )
+            print(result)
```
Comment (taylorfturner) [inline tests/distinct_generators/test_categorical_generator.py:40]: remove

**Review by taylorfturner — CHANGES_REQUESTED** [review]

File: synthetic_data/distinct_generators/categorical_generator.py [hunk 6]
```diff
@@ -17,11 +20,18 @@ def random_categorical(
     :type categories: string, None, optional
     :param num_rows: the number of rows in np array generated
     :type num_rows: int, optional
+    :param p: a list of floats that respresent the probability each category will be chosen
+    :type p: float, optional
```
Comment (taylorfturner) [inline synthetic_data/distinct_generators/categorical_generator.py]: issue here .... should be `probabilities` not `p`

**Review by micdavis — APPROVED** [review]

**Review by taylorfturner — DISMISSED** [review]

**Review by JGSweets — APPROVED** [review]

**Review by taylorfturner — APPROVED** [review]

**Pull request merged**